### ISSUE TITLE
No CI/CD for the project.

### ISSUE THREAD

**Issue by jwittner** [GlitchEnzo/NuGetForUnity#113]
@GlitchEnzo I'm tying to setup a CI/CD process for this for our internal usage but I'm running into this issue:

> Cannot get the webhooks of 'https://api.github.com/repos/GlitchEnzo/NuGetForUnity' from Github. This is likely caused by not having the necessary permission to manage hooks for the selected repository.

Think you can give me permissions to hook into this repo so we can directly create builds? I'd be open to a conversation about using Travis or something public, especially if we can get a unitypackage deployed to the Asset store.

Thoughts?

**Comment by GlitchEnzo** [comment]
I don't see any way to give you permission to do that without transferring ownership.  Are you aware of a way to give a contributor permission to webhooks?

I can add webhooks that you need, though.

**Comment by jwittner** [comment]
Looks like in organizations you might be able to achieve this with a team, but it doesn't look possible outside of that. For now I think we'll run our CI against my fork.

Related to your long term, how do you feel about setting some CI/CD up and getting this on the Unity Asset store? If you're interested, I'd love to help in any way I can.

**Comment by GlitchEnzo** [comment]
I'm definitely interested in setting up some CI/CD, and I've always wanted to put NuGetForUnity on the Unity Asset Store once all of the major issues have been resolved.

It'd be nice to add more unit tests and have them run on various versions and configurations (.NET 3.5 vs .NET 4.6) of Unity.

**Comment by jwittner** [comment]
I've never used Travis CI, but have heard it's free for OSS projects. I found at least one person claiming to build Unity projects with it: https://jonathan.porta.codes/2015/04/17/automatically-build-your-unity3d-project-in-the-cloud-using-travisci-for-free/

**Comment by GlitchEnzo** [comment]
I've started working on setting up Travis CI:
https://github.com/GlitchEnzo/NuGetForUnity/tree/travis

**Comment by jwittner** [comment]
Retitled this to encompass CI/CD generally. =)

**Comment by GlitchEnzo** [comment]
I see the build server having 3 major steps:
1. Build the NugetForUnity DLL
2. Build the .unitypackage containing the DLL
3. Upload the .unitypackage somewhere (GitHub or Asset Store)

In order to build the NuGetForUnity DLL, we need the UnityEngine and UnityEditor DLLs. I'm currently grabbing them out of the Library folder from the project.  This requires that the project be opened by Unity and have the assets "imported" into it, thus creating the Library folder with Unity's DLLs.

Due to Travis having a fresh, clean environment for every build (which is a good thing), each build requires downloading and installing Unity from scratch.  This can bloat up the build time ~15 minutes as it downloads the 2 GB installer. I'd prefer to avoid this, especially for the first step of simply building our DLL, which is the most important step to get fast feedback on.

Unfortunately, there are no NuGet packages that contain the Unity DLLs. ([This one](https://www.nuget.org/packages/Unity3D.DLLs/) simply has PowerShell scripts to point to your installed DLLs.) So I'm debating creating my own UnityEditor and UnityEngine NuGet packages with the DLLs, or simply checking the 5.2 DLLs into the repo so we can compile the DLL without a Unity installation.

Building the .unitypackage would also require an installation of Unity (unless we write own packager since it's a [known](https://www.gamedev.net/forums/topic/659386-any-interest-in-reverse-engineering-unitypackage-files/) [file format](https://github.com/gered/extractunitypackage/blob/master/extractunitypackage.py)), but it would happen _after_ the main build.

Does anyone else have any preferences? Create NuGet packages or check them in? Something else entirely?

**Comment by jwittner** [comment]
My only concern with committing to this repo is if distributing the DLLs violates any of Unity's terms, if we're in the clear I'd vote for doing it to move this forward =).  Out of curiosity, are you using chocolatey to install Unity? We recently discovered the Unity packages there.

**Comment by GlitchEnzo** [comment]
It appears that the Mac VM on Travis doesn't have a C# compiler installed by default.  Mono would have to be downloaded and installed to even compile the code, so I'm thinking we install Unity as originally planned. We can use the DLLs distributed by Unity and the Mono C# compiler included with it.

> Out of curiosity, are you using chocolatey to install Unity? 

I am not. I'm using curl to download the installer directly from Unity's site:
https://github.com/GlitchEnzo/NuGetForUnity/blob/travis/BuildScripts/install_unity.sh

**Comment by GlitchEnzo** [comment]
I decided to move away from Travis. Travis only supports Mac and Linux and I'm not as familiar with either of those as I am with Windows.  

AppVeyor is Windows-only, but is otherwise very similar to Travis. I hooked up this repo to AppVeyor and it automatically found the CreateDLL.sln file and tried building it with MSBuild.  It was missing the Unity DLLs, so I used PowerShell to download and install Unity.  Now, it is successfully building NuGetForUnity.dll.

https://ci.appveyor.com/project/GlitchEnzo/nugetforunity

I even added in an additional PowerShell script to create the .unitypackage file, too.  It is added as a build artifact:
https://ci.appveyor.com/project/GlitchEnzo/nugetforunity/build/artifacts
